content/blog/2020/03/2020-03-24-azure-key-vault-cred-provider.adoc

```diff
@@ -0,0 +1,197 @@
+
+1. click 'Credentials'
+
+2. click 'System' (it'll appear below the Credentials link in the side bar)
+
+3. click 'Global credentials (unrestricted)'
+
+4. click 'Add Credentials'
+
+5. select 'Microsoft Azure Service Principal'
+image:/images/post-images/azure-key-vault/azure-service-principal-drop-down.png[Microsoft Azure Service Principal dropdown]
+
+6. fill out the form from the credential created above, appId is 'Client ID', password is 'Client Secret'
+image:/images/post-images/azure-key-vault/azure-service-principal-configuration.png[Microsoft Azure Service Principal credential configuration]
+
+7. click 'Verify Service Principal', you should see 'Successfully verified the Microsoft Azure Service Principal'.
+
+8. click 'Save'
+
+=== Jenkins Azure Key Vault plugin configuration
+
+You now have a credential you can use to interact with Azure resources from Jenkins, now you need to configure the plugin:
+
+1. go back to the Jenkins home page
+
+2. click 'Manage Jenkins'
+
+3. click 'Configure System'
+
+4. search for 'Azure Key Vault Plugin'
+
+5. enter your vault url and select your credential
+image:/images/post-images/azure-key-vault/azure-key-vault-global-config.png[Azure Key Vault plugin configuration]
+
+6. click 'Save'
+
+=== Store a secret in Azure Key Vault
+
+For the step after this you will need a secret, so let's create one now:
+
+[source, shell]
+----
+$ az keyvault secret set --vault-name $YOUR_VAULT --name secret-key --value my-super-secret
+----
+
+=== Create a pipeline
+
+Install the link:https://plugins.jenkins.io/workflow-aggregator/[Pipeline] plugin if you don't already have it.
+
+From the Jenkins home page, click 'New item', and then:
+
+1. enter a name, i.e. 'key-vault-test'
+
+2. click on 'Pipeline'
+
+3. add the following to the pipeline definition:
+
+[pipeline]
+----
+// Declarative //
+pipeline {
+  agent any
+  environment {
+    SECRET_KEY = credentials('secret-key')
+  }
+  stages {
+    stage('Foo') {
+      steps {
+        echo SECRET_KEY
+        echo SECRET_KEY.substring(0, SECRET_KEY.size() - 1) // shows the right secret was loaded, don't do this for real secrets unless you're debugging
+      }
+    }
+  }
+}
+
+// Scripted //
+withCredentials([string(credentialsId: 'secret-key', variable: 'SECRET_KEY')]) {
+    echo SECRET_KEY
+    echo SECRET_KEY.substring(0, SECRET_KEY.size() - 1) // shows the right secret was loaded, don't do this for real secrets unless you're debugging
+}
+----
+
+You have now successfully retrieved a credential from Azure Key Vault using native Jenkins credentials integration.
+
+== configuration-as-code integration
+
+[.lead]
+The Configuration as Code plugin has been designed as an opinionated way to configure Jenkins based on human-readable declarative configuration files. Writing such a file should be feasible without being a Jenkins expert, just translating into code a configuration process one is used to executing in the web UI.
```

Comment:
`just translating into code a configuration process one is used to executing in the web UI` just looks wired to me.